### DISCUSSION TITLE
some chat_receive research

### DISCUSSION THREAD

**Discussion by anzz1** [YimMenu/YimMenu#536] [Research]
Some research I made on the chat_receive func at 4D 85 C9 0F 84 ? ? ? ? 48 8B C4 48 89 58 08 48 89 70 10 48 89 78 18 4C 89 48 20 for a simple antispam hook may be of some use so I'll just paste the post here:

> Here is some info for the people interested in the reversing part:
> 
> So it seems that indeed some compiler optimization is involved and the type of chat_receive function is probably __vectorcall instead of __fastcall, as a vector argument is used in the shadow space of the stack. IDA doesn't seem to quite understand vectorcalls so I couldn't make it produce clean output.
> 
> Here is the results of my preliminary reversing, the function could look something like this. Note that a lot of guesswork is involved so don't assume any of it is actually right.
> 
> ```C
> void __fastcall chat_receive(
>         struct unk_chat *this,
>         _QWORD *a2,
>         rage::rlGamerHandle *ghSender,
>         char *msg,
>         char clan_member_id)
> {
>   bool v7; // r13
>   rage::netPlayer *netPlayer_0; // rsi
>   uint64_t v9; // rax
>   int v10; // edi
>   rage::rlGamerInfo *gamerInfos; // rbx
>   int i; // er15
>   int v13; // ebx
>   unsigned int v14; // er15
>   rage::rlGamerHandle *gamerHandle_1; // rax
>   rage::rlGamerInfo *gamerInfo; // r12
>   Concurrency::details::ScheduleGroupSegmentBase *groupSegmentBase; // rbx
>   rage::rlGamerHandle *gamerHandle; // rax
>   __int64 v19; // rax
>   __int64 clanid_low; // rdx
>   _BYTE *v21; // rax
>   char *v22; // rdx
>   int v23; // ebx
>   int v24; // eax
>   rage::netPlayer *v25; // rbx
>   int j; // edi
>   unk_chat unkChat; // [rsp+20h] [rbp-E0h] BYREF
>   char v28; // [rsp+1388h] [rbp+1288h] BYREF
> 
>   if ( msg )
>   {
>     v7 = *(_DWORD *)(qword_14262FA70 + 110392) == 4;
>     netPlayer_0 = sub_140C57494(ghSender, (__int64)a2);
>     v9 = sub_140CC78D8((__int64)off_141AF74F8);
>     v10 = 0;
>     *(_QWORD *)&unkChat.field_40 = v9;
>     if ( netPlayer_0 || v7 )
>     {
>       if ( v9 )
>       {
>         gamerInfos = unkChat.gamerInfos;
>         for ( i = 31; i >= 0; --i )
>           rage::rlGamerInfo::ctor(gamerInfos++);
>         if ( v7 )
>         {
>           v13 = 0;
>           v14 = sub_140D462D0(qword_14262FA70);
>           if ( !v14 )
>             goto LABEL_41;
>           while ( 1 )
>           {
>             gamerHandle_1 = rage::rlGamerInfo::get_gamer_handle(&unkChat.gamerInfos[v13]);
>             if ( rage::rlGamerHandle::Equal(gamerHandle_1, ghSender) )
>               break;
>             if ( ++v13 >= v14 )
>               goto LABEL_41;
>           }
>           if ( v13 < 0 )
>             goto LABEL_41;
>           gamerInfo = &unkChat.gamerInfos[v13];
>         }
>         else
>         {
>           gamerInfo = (rage::rlGamerInfo *)((__int64 (__fastcall *)(rage::netPlayer *))netPlayer_0->get_net_data)(netPlayer_0);
>         }
>         groupSegmentBase = off_141AF7528;
>         gamerHandle = rage::rlGamerInfo::get_gamer_handle(gamerInfo);
>         if ( sub_140D4C658(groupSegmentBase, gamerHandle)
>           || !sub_13FD5D1A8((__m128i *)this->clan_member_id_2.m128i_i32, (__int64)"ADD_MESSAGE") )
>         {
> LABEL_41:
>           v25 = (rage::netPlayer *)&v28;
>           for ( j = 31; j >= 0; --j )
>           {
>             v25 = CONTAINING_RECORD(v25, rage::netPlayer, pad_0098);
>             sub_140FF6E84(v25);
>           }
>           return;
>         }
>         v19 = sub_140FC27D4(gamerInfo);
>         sub_13FD59CCC((__int64)this->clan_member_id_2.m128i_i64, v19, 1);
>         sub_13FD59CCC((__int64)this->clan_member_id_2.m128i_i64, (__int64)msg, 1);
>         if ( clan_member_id )
>         {
>           clanid_low = LODWORD(this->clanid);
>           if ( (_DWORD)clanid_low )
>           {
>             v21 = (_BYTE *)sub_1409868BC(&unk_1423AB150, clanid_low, "MP_CHAT_TEAM");
>             goto LABEL_23;
>           }
>           v22 = "MP_CHAT_TEAM";
>         }
>         else
>         {
>           v22 = "MP_CHAT_ALL";
>         }
>         v21 = sub_140986904((__int64)&unk_1423AB150, v22);
> LABEL_23:
>         sub_13FD59CCC((__int64)this->clan_member_id_2.m128i_i64, (__int64)v21, 1);
>         unkChat.clan_member_id.m128i_i8[0] = clan_member_id;
>         unkChat.clan_member_id.m128i_i32[2] = 2;
>         unkChat.clan_member_id_2 = unkChat.clan_member_id;
>         unkChat.clanid_2 = unkChat.clanid;
>         sub_13FD59624((__int64)this->clan_member_id_2.m128i_i64, (__m128i *)unkChat.clan_member_id_2.m128i_i64);
>         if ( !(unsigned __int8)sub_140C61568() )
>         {
>           v10 = 1;
> LABEL_38:
>           unkChat.clan_member_id.m128i_i32[2] = 1;
>           unkChat.clanid_2 = unkChat.clanid;
>           *(double *)unkChat.clan_member_id.m128i_i64 = (double)v10;
>           unkChat.clan_member_id_2 = unkChat.clan_member_id;
>           sub_13FD59624((__int64)this->clan_member_id_2.m128i_i64, (__m128i *)unkChat.clan_member_id_2.m128i_i64);
>           sub_14065C21C(&this->clan_member_id_2);
>           if ( !BYTE4(this->clanid) && !this->clan_member_id.m128i_i32[1] )
>           {
>             sub_13FEDC60C(this, 1, 0);
>             this->clan_member_id.m128i_i32[0] = Seed;
>           }
>           goto LABEL_41;
>         }
>         if ( v7 )
>         {
>           if ( !netPlayer_0 )
>             goto LABEL_35;
>         }
>         else
>         {
>           if ( !netPlayer_0 )
>             goto LABEL_38;
>           if ( BYTE5(this->clanid) )
>           {
>             v23 = sub_14131D1FC(*(_QWORD *)&unkChat.field_40);
>             if ( (unsigned int)sub_14131D1FC(netPlayer_0) == v23 )
>             {
> LABEL_32:
>               v10 = 118;
>               goto LABEL_38;
>             }
>             v24 = sub_14131D1FC(netPlayer_0) - 1;
>             if ( !v24 )
>             {
>               v10 = 29;
>               goto LABEL_38;
>             }
>             if ( v24 == 1 )
>             {
>               v10 = 30;
>               goto LABEL_38;
>             }
> LABEL_35:
>             v10 = 28;
>             goto LABEL_38;
>           }
>         }
>         if ( sub_1412755A4((__int64)netPlayer_0) )
>           goto LABEL_32;
>         goto LABEL_35;
>       }
>     }
>   }
> }
> ```
> 
> The garbage under LABEL_23 and LABEL_38 is the result of the __vectorcall argument, it is located at rsp+18, in the shadow stack space, which I think is of type __m128i and also the first field of unk_chat struct. I couldn't force IDA to align the unkChat struct to +18 so it's at normal first arg pos at +20 and probably misaligned. 
> 
> The "clanids" are probably crew and/or organization stuff.
> 
> The chat_receive itself is probably a virtual function inside a class, something like rage::Network::ChatManager chatMgr->chat_receive()
> 
> The arguments are probably as follows:
> 1 : (class*) pointer to the class (this), hidden
> 2 : (uint64_t) sender peer_id / host_token / something.
> 3 : (rage::rlGamerHandle*) pointer to sender gamer handle
> 4 : (const char*) message
> 
> 5 : the special case, probably a compiler optimization and a vector argument where the first 64 bits of the arg is located in the "shadow stack" at +18 and 2nd 64-bits at the regular 1st stack-arg position (+20). This is what confuses the hex-rays decompiler. It looks like to be type __m128i and also probably the first field of the chatMgr struct. 
> 
> ---
> 
> So this is where I first started investigating the crash, I thought that it's a stack misalignment issue because the simple way of hooking, inserting our function in between changes the stack position and maybe the func needs to access something further down the stack, position of which is now different than before so offsets would point to wrong places. Also by not using the __vectorcall type the vector arg at shadow stack space is now garbage data. The function also moves the stack position quite a bit, 4796 bytes, so adding our func in between could just maybe result in a stack overflow. Also it might be possible that the stack is aligned in a very specific way where the variables IDA thinks are "arguments" are actually part of the "chatManager" class and the stack is aligned so that the class accesses its' own fields. So changing the stack position would lead to wrong accesses and crash. There is also the potential of some function further down the line throwing an exception, but as after adding our func the stack unwind data is wrong, the exception won't be caught properly and stack will be unwound to garbage result.
> 
> So even with a ton of more research to know everything, fixing the argument types and the function itself and fixing the unwind data in memory would still end up having the stack in a different place and not being sure whether there is enough space on the stack.
> 
> So I decided that the best course of action is that instead of doing all that, just simply rewrite the hook. The old hook simply made my_chat_receive func go in between the caller func and callee func, thus remaining in the call stack with wrong unwind info and have different stack position.
> 
> The new hook is written in assembly (instead of trying to fight the compiler to make it do what i want, not to mention different results from different compilers) and it completely separates the my_chat_receive func from the control flow. It saves the original arguments and stack state, calls our func, restores the arguments and stack and then either returns to skip the message or continues with the original function. Doing it this way makes it as if our hook/function never even was there so it don't have any of the caveats mentioned before. Stack will stay at its' original position, arguments stay exactly as they were (irregardless of their type) and exception handling / stack unwinding will continue to work as before.
> 
> ```C
> /*
>  *  push rcx                  // save arg1
>  *  push rdx                  // save arg2
>  *  push r8                   // save arg3
>  *  sub rsp,0x28              // move stack by 40 bytes
>  *  mov rcx, r9               // arg4 -> arg1
>  *  lea r11, 00000000         // [rip+disp32]
>  *  call r11                  // call my_chat_receive
>  *  add rsp,0x28              // restore stack
>  *  pop r8                    // restore arg3
>  *  pop rdx                   // restore arg2
>  *  pop rcx                   // restore arg1
>  *  test al,al                // check return from my_chat_receive
>  *  jz [rip+0x1A]             // jump to ret if zero
>  *  mov r9, rax               // restore arg4
>  *  mov rax, rsp              // orig function stub
>  *  mov [rax+8], rbx          // orig function stub
>  *  jmp QWORD PTR [rip+0x0]   // jmp to ret_jmp_ptr and don't return
>  *  0000000000000000          // ret_jmp_ptr addr
>  *  ret                       // skip message
>  *
>  */
> static BYTE m_chat_receive_stub[] = {
>   0x51,0x52,0x41,0x50,0x48,0x83,0xEC,0x28,0x4C,0x89,0xC9,0x4C,0x8D,0x1D,0,0,0,0,0x41,0xFF,
>   0xD3,0x48,0x83,0xC4,0x28,0x41,0x58,0x5A,0x59,0x84,0xC0,0x74,0x18,0x49,0x89,0xC1,0x48,0x89,0xE0,0x48,
>   0x89,0x58,0x08,0xFF,0x25,0,0,0,0,0,0,0,0,0,0,0,0,0xC3
> };
> ```
> 
> After all this, I realized what made it crash was actually none of this but a simple and stupid error on my part. It was caused by having the last_msg[] array on the same memory page with the chat_receive_stub, and by setting the protection to PAGE_EXECUTE_READ, it made last_msg[] lose write access and thus writing to it obviously resulted in a access violation. :facepalm:
> 
> So now the access rights are set properly, adding only what's required and not removing any existing rights.
> 
> That simple change fixed the crash. In any case, the new way of hooking is still way better than the old one and should be 100% stable and non-obtrusive. 
> 
> I probably won't be pursuing this any further so I hope the write-up and research helps someone who is interested in reversing the chat functionality.
> 

Post : https://www.unknowncheats.me/forum/grand-theft-auto-v/518815-gta-online-unplayable-nowadays-2.html
Antispam hook : https://github.com/anzz1/gta-antispam

**Comment by maybegreat48** [comment]
This is some very good research. You can also go up the callstack and intercept chat directly from the message handler
```c++
else if (msg_type == MessageType::MsgTextMessage ||
				msg_type == MessageType::MsgTextMessage2)
{
	char buf[0x100]{};
	if (player && buffer.ReadString(buf, 0x100))
	{
		if (player->m_is_spammer)
			return true;

		if (spam::is_text_spam(buf))
		{
			player->m_is_spammer = true;
			spam::log_chat(buf, player, true);
			return true;
		}
		else
		{
			spam::log_chat(buf, player, false);
		}
	}
}
```
